pyo3-macros-backend/src/pymethod.rs

```diff
@@ -409,14 +414,17 @@ fn impl_traverse_slot(cls: &syn::Type, rust_fn_ident: &syn::Ident) -> MethodAndS
         ) -> ::std::os::raw::c_int
         {
             let trap = _pyo3::impl_::panic::PanicTrap::new("uncaught panic inside __traverse__ handler");
-            let pool = _pyo3::GILPool::new();
-            let py = pool.python();
+            let py = _pyo3::Python::assume_gil_acquired();
             let slf = py.from_borrowed_ptr::<_pyo3::PyCell<#cls>>(slf);
-
-            let visit = _pyo3::class::gc::PyVisit::from_raw(visit, arg, py);
             let borrow = slf.try_borrow();
+            let visit = _pyo3::class::gc::PyVisit::from_raw(visit, arg, py);
+
             let retval = if let ::std::result::Result::Ok(borrow) = borrow {
-                _pyo3::impl_::pymethods::unwrap_traverse_result(borrow.#rust_fn_ident(visit))
+                let _lock = _pyo3::impl_::LockGIL::new();
+                match ::std::panic::catch_unwind(::std::panic::AssertUnwindSafe(move || borrow.#rust_fn_ident(visit))) {
```

Comment:
I am not sure if we want to the `catch_unwind` dance here? It does seem necessary to make this testable as the `PanicTrap` aborts the process otherwise. But what is the desired behaviour in production code? Is `tp_traverse` returning -1 sufficient or do we actually want to abort the process when this happens?

(I also wonder whether this dance should be encapsulated better, e.g. replace `unwrap_traverse_result` by `call_traverse_impl` which does that and creates the `LockGIL` instance and handles any resulting panic.)

Reply:
(Added a commit which moves as much of this as possible into a library function instead of expanding it in the macros.)

Reply:
After thinking about this some more, we definitely do not want to abort the process because something accidentally (and possibly indirectly) call `with_gil`. Furthermore, the `catch_unwind` dance is the same we do for the other trampolines.